src/HttpClientFactory/HeaderPropagation/src/HeaderPropagationMessageHandler.cs

```diff
@@ -0,0 +1,54 @@
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Http.HeaderPropagation
+{
+    public class HeaderPropagationMessageHandler : DelegatingHandler
+    {
+        private readonly HeaderPropagationOptions _options;
+        private readonly IHttpContextAccessor _contextAccessor;
+
+        public HeaderPropagationMessageHandler(HeaderPropagationOptions options, IHttpContextAccessor contextAccessor)
+        {
+            _options = options;
+            _contextAccessor = contextAccessor;
+        }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (_contextAccessor.HttpContext != null)
+            {
+                foreach (var header in _options.Headers)
+                {
+                    if (!_contextAccessor.HttpContext.Request.Headers.TryGetValue(header.InputName, out var values)
```

Comment:
@davidfowl we've run into trouble trying to use IHttpContextAccessor in DelegatingHandlers before haven't we? People would run their HttpClient requests in parallel and cause unsafe mulithreaded access to HttpContext.  When ApplicationInsights dealt with a similar problem they concluded the best way to handle it was to collect fields they needed up front and pass them through in a dedicated data structure.

What if instead of a DelegatingHandler we provided a set of request composition helpers on top of HttpClient and HttpRequestMessage?

The biggest downside I can see would be the caller needing to be HttpContext aware, which may or may not work for various libraries.

Reply:
Hi Tratcher,
yes the HttpContext is not thread-safe and it's a really valid point, not wure why I didn't considered sending requests in parallel :(
I should add a `HeaderPropagationInitializer` similar to the `TelemetryInitializer` that get the required values from the context and store it in it's own data structure. However this will mean that it needs to be aware of all the headers forwarded by all the `HttpClient`s as we will want to lock on the context only once per MVC request. I'll have to think about it.

About the composition helpers, I'm not sure how that will fit with the `HttpClientFactory`. Can you elaborate a bit more on what you have in mind?

Btw here is the issue you are referring to:
https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues/373
I remembered he mentioned it on last year talk at NDC :)

Reply:
The composition helpers wouldn't be affiliated with HttpClientFactory, they'd be APIs for directly creating or operating on HttpRequestMessages before calling into HttpClient. They may have extensions on HttpClient to facilitate.

Reply:
Yeah, I think the right thing to do here is to make a middleware that initializes state, and a standalone `DelegatingHandler` that reads that state to modify the outgoing request.

I don't think we need to tie this to the http client factory. In general library code doesn't need to integrate with it directly because wiring up a `DelegatingHandler` is easy.